Associations

Alejandro-Bernal-M opened the pull request:
# Creating the association between classes:
<img src="https://media4.giphy.com/media/OijlHruJk544wGchMP/giphy.gif"/>

## Add the following changes:

- Create a class Classroom with the following:
     - label instance variable, should be initialized in the constructor.
     - Setter and getter for label.
- Create the has-many/belongs-to relationship between Classroom and Student. Implemented:
    - the has-many side (a classroom has many students).
     - the belongs-to side (a student belongs to a classroom).
- Create a class Book with the following:
 - title and author instance variables, initialized in the constructor.
  - Setters and getters for instance variables.
- Create a class Rental with the following:
     - date instance variable, initialized in the constructor.
     - Setter and getter for date.
 - Create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. Implemented:
    - Create the has-many side of Book and Rental (a book has many rentals).
    - Create the belongs-to side of Rental and Book (a rental belongs to a book).
    - Create the has-many side of Person and Rental (a person has many rentals).
    - Create the belongs-to side of Rental and Person (a rental belongs to a person).
    - Modify the constructor of Rental so Book and Person are set in it.

Alejandro-Bernal-M committed:
Create classroom class

Alejandro-Bernal-M committed:
Add has to many association to classroom

Alejandro-Bernal-M committed:
Add belongs to relationship on student and correct calling on students and classroom

Alejandro-Bernal-M committed:
Add book class

Alejandro-Bernal-M committed:
Add rental class

Alejandro-Bernal-M committed:
Add many to many relationship between book and person through rental class

christianbiring1 commented on classroom.rb:
```diff
@@ -0,0 +1,14 @@
+    @students = []
+  end
+
+  def add_student(student)
+    @students.push(student)
+    student.classroom = self
```
- [x] Great work on implementing the `has_many` relationship between the students and classroom :+1:

christianbiring1 commented on student.rb:
```diff
@@ -1,11 +1,18 @@
     @classroom = classroom
   end
 
+  def classroom=(classroom)
+    @classroom = classroom
+    classroom.students.push(self) unless classroom.students.include?(self)
+  end
```
- [x] Great work on implementing the `belongs_to` relationship between students and class :+1:

christianbiring1 commented on book.rb:
```diff
@@ -0,0 +1,10 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
```
[OPTIONAL]
- [x] Kindly make sure the `rentals` are also in the `setter` 
```rb
attr_accessor
```
As it can be modified and not read-only

christianbiring1 commented on rental.rb:
```diff
@@ -0,0 +1,11 @@
+class Rental
+  attr_accessor :date
+
+  def initialize(date, book, student)
+    @date = date
+    @book = book
+    book.visits.push(self)
+    @student = student
+    student.visits.push(self)
+  end
```
- [x] As specified in the project requirements, Kindly make sure a rental `belongs` to a `book` and a `person` and not a student as you just tried to do.
Hint: You can achieve this by saying 
```rb
def initialize(date, book, person)
# *The rental is adding itself to the @book achieving the 'belongs-to' association
    book.rentals << self

    @person = person
    # *The rental is adding itself to the @person the 'belongs-to' association
    person.rentals << self
end
```

christianbiring1 commented on person.rb:
```diff
@@ -4,14 +4,15 @@
     @parent_permission = parent_permission
     super()
+    @rentals = []
   end
```
- [ ] Kinldy makes sure a person has `many` rentals as specified in the project requirements.

Hint: You may consider saying something like :
```rb
def add_rental(book, date)
    # *The person itself should be sent as a parameter to create the new rental achieving the 'has-many' association
    Rental.new(date, book, self)
  end
```

christianbiring1 reviewed (changes requested):
Hi @Alejandro-Bernal-M , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
# STATUS: REQUEST CHANGES :recycle: 
![soclose](https://github.com/Alejandro-Bernal-M/school_library/assets/97100091/69ffa135-93f5-4648-827e-86ba687027fc)

## Highlights
- [ ] Great work for implementing the `belongs_to` `many _to_many` between classroom and student :+1: 
- [ ] Descriptive PR `title` and `summary` :+1: 
- [ ] All linters checks are passing :+1: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- x
- y
- z

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Alejandro-Bernal-M committed:
Corrections

Alejandro-Bernal-M commented:
Hello, I corrected my mistakes, but I implemented some things differently as you propose, following the recommendations of Microverse https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md

I created a new file called main.rb where I wrote some cases, these are the results:
![image](https://github.com/Alejandro-Bernal-M/school_library/assets/118059775/1873f87e-809d-43f8-aec2-42842f164e9d)

prolajumokeoni commented on person.rb:
```diff
@@ -33,9 +34,10 @@ def can_use_services?
-puts capitalized_person.correct_name
-capitalized_trimmed_person = TrimmerDecorator.new(capitalized_person)
-puts capitalized_trimmed_person.correct_name
+# tests for decorators =>
+# person = Person.new(22, 'maximilianus')
+# person.correct_name
+# capitalized_person = CapitalizeDecorator.new(person)
+# puts capitalized_person.correct_name
+# capitalized_trimmed_person = TrimmerDecorator.new(capitalized_person)
+# puts capitalized_trimmed_person.correct_name
```
## OPTIONAL 

- [ ] Kindly remove all commented code from your project, you can leave comments about your code but not commented code, this will make your codebase look neat and readable 😉

prolajumokeoni reviewed (approved):
Hi @Alejandro-Bernal-M 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
## Highlights
- Associations setup correctly

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

_Check inline comment_

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification @prolajumokeoni.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
